tests/contracts/oas20-wrapped/contract/src/lib.rs

```diff
@@ -0,0 +1,636 @@
+//! Oasis Wrapped-OAS20 contract, to be used by the Oasis Wormhole Token Bridge.
+//!
+//! This is based on the `contract-sdk/specs/oas20` contract with added support
+//! for minter/owner to burn tokens from any account.
+#![feature(wasm_abi)]
+
+extern crate alloc;
+
+use oasis_contract_sdk::{self as sdk, types::message::NotifyReply};
+use oasis_contract_sdk_oas20_types as oas20;
+use oasis_contract_sdk_storage::{cell::Cell, map::Map};
+use oasis_contract_sdk_types::address::Address;
+
+use oasis_oas20_wrapped_types as types;
+use types::{Error, Event, Request, Response};
+
+/// The contract type.
+pub struct Oas20Wrapped;
+
+/// Storage cell for the token information.
+const TOKEN_INFO: Cell<oas20::TokenInformation> = Cell::new(b"token_info");
+/// Storage map for account balances.
+const BALANCES: Map<Address, u128> = Map::new(b"balances");
+/// Storage map for allowances.
+const ALLOWANCES: Map<(Address, Address), u128> = Map::new(b"allowances");
+/// Storage cell for wormhole bridge wrapped token information.
+const WRAPPED_INFO: Cell<types::BridgeWrappedInfo> = Cell::new(b"wrapped_info");
+
+impl sdk::Contract for Oas20Wrapped {
+    type Request = Request;
+    type Response = Response;
+    type Error = Error;
+
+    fn instantiate<C: sdk::Context>(ctx: &mut C, request: Request) -> Result<(), Error> {
+        match request {
+            Request::Instantiate {
+                token_instantiation,
+                asset_chain_id,
+                asset_address,
+            } => {
+                // Ensure an owner (minter) is set, otherwise a wrapped token doesn't make sense.
+                if token_instantiation.minting.is_none() {
+                    return Err(Error::MinterNotConfigured);
+                }
+
+                let token_information =
+                    oas20::helpers::instantiate(ctx, BALANCES, TOKEN_INFO, token_instantiation)?;
+
+                let wrapped_info = types::BridgeWrappedInfo {
+                    asset_chain_id,
+                    asset_address,
+                    bridge_address: ctx.caller_address().to_owned(),
+                };
+                WRAPPED_INFO.set(ctx.public_store(), wrapped_info.clone());
+
+                ctx.emit_event(Event::WrappedOas20Instantiated {
+                    token_information,
+                    wrapped_info,
+                });
+
+                Ok(())
+            }
+            _ => Err(Error::BadRequest),
+        }
+    }
+
+    fn call<C: sdk::Context>(ctx: &mut C, request: Request) -> Result<Response, Error> {
+        // This method is called for each contracts.Call call. It is supposed to handle the request
+        // and return a response.
+        match request {
+            // Handle basic Oas20 requests.
+            Request::Oas20(request) => {
```

Comment:
Are these requests all standard? Could we have a helper that can handle all of these `Oas20` requests to avoid reimplementations (just passing in all the storage cells or something)?